File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/network/EntityTrackerServer.java
package net.namekdev.entity_tracker.network;

import java.net.SocketAddress;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;

import com.artemis.utils.Bag;
import com.artemis.utils.BitVector;
import net.namekdev.entity_tracker.connectors.WorldController;
import net.namekdev.entity_tracker.connectors.WorldUpdateListener;
import net.namekdev.entity_tracker.model.AspectInfo;
import net.namekdev.entity_tracker.model.ComponentTypeInfo;
import net.namekdev.entity_tracker.network.base.RawConnectionCommunicator;
import net.namekdev.entity_tracker.network.base.RawConnectionCommunicatorProvider;
import net.namekdev.entity_tracker.network.base.RawConnectionOutputListener;
import net.namekdev.entity_tracker.network.base.Server;
import net.namekdev.entity_tracker.network.communicator.EntityTrackerCommunicator;
import net.namekdev.entity_tracker.utils.tuple.Tuple3;

/**
 * Server listening to new clients, useful to pass into Entity Tracker itself.
 * Collects data to gather world state for incoming connections.
 *
 * @author Namek
 */
public class EntityTrackerServer extends Server implements WorldUpdateListener {
	private WorldController _worldController;
	private Bag<EntityTrackerCommunicator> _listeners = new Bag<EntityTrackerCommunicator>();

	private Bag<String> _managers = new Bag<String>();
	private Bag<Tuple3<Integer, String, AspectInfo>> _systems = new Bag<Tuple3<Integer, String, AspectInfo>>();
	private Bag<ComponentTypeInfo> _componentTypes = new Bag<ComponentTypeInfo>();
	private Map<Integer, BitVector> _entities = new HashMap<>();
	private Bag<Integer> _entitySystemsEntitiesCount = new Bag<Integer>();
	private Bag<Integer> _entitySystemsMaxEntitiesCount = new Bag<Integer>();


	public EntityTrackerServer() {
		this(DEFAULT_PORT);
	}

	public EntityTrackerServer(int listeningPort) {
		super();
		super.clientListenerProvider = _communicatorProvider;
		super.listeningPort = listeningPort;
	}

	@Override
	public void injectWorldController(WorldController controller) {
		_worldController = controller;
	}

	@Override
	public int getListeningBitset() {
		// TODO
		return ENTITY_ADDED | ENTITY_DELETED | ENTITY_SYSTEM_STATS;
	}

	@Override
	public void addedSystem(int index, String name, BitVector allTypes, BitVector oneTypes, BitVector notTypes) {
		for (int i = 0, n = _listeners.size(); i < n; ++i) {
			EntityTrackerCommunicator communicator = _listeners.get(i);
			communicator.addedSystem(index, name, allTypes, oneTypes, notTypes);
		}
		_systems.add(Tuple3.create(index, name, new AspectInfo(allTypes, oneTypes, notTypes)));
	}

	@Override
	public void addedManager(String name) {
		for (int i = 0, n = _listeners.size(); i < n; ++i) {
			EntityTrackerCommunicator communicator = _listeners.get(i);
			communicator.addedManager(name);
		}
		_managers.add(name);
	}

	@Override
	public void addedComponentType(int index, ComponentTypeInfo info) {
		for (int i = 0, n = _listeners.size(); i < n; ++i) {
			EntityTrackerCommunicator communicator = _listeners.get(i);
			communicator.addedComponentType(index, info);
		}
		_componentTypes.set(index, info);
	}

	@Override
	public void updatedEntitySystem(int systemIndex, int entitiesCount, int maxEntitiesCount) {
		for (int i = 0, n = _listeners.size(); i < n; ++i) {
			EntityTrackerCommunicator communicator = _listeners.get(i);
			communicator.updatedEntitySystem(systemIndex, entitiesCount, maxEntitiesCount);
		}
		_entitySystemsEntitiesCount.set(systemIndex, entitiesCount);
		_entitySystemsMaxEntitiesCount.set(systemIndex, maxEntitiesCount);
	}

	@Override
	public void addedEntity(int entityId, BitVector components) {
		for (int i = 0, n = _listeners.size(); i < n; ++i) {
			EntityTrackerCommunicator communicator = _listeners.get(i);
			communicator.addedEntity(entityId, components);
		}
		_entities.put(entityId, components);
	}

	@Override
	public void deletedEntity(int entityId) {
		for (int i = 0, n = _listeners.size(); i < n; ++i) {
			EntityTrackerCommunicator communicator = _listeners.get(i);
			communicator.deletedEntity(entityId);
		}
		_entities.remove(entityId);
	}

	@Override
	public void updatedComponentState(int entityId, int componentIndex, Object[] values) {
		for (int i = 0, n = _listeners.size(); i < n; ++i) {
			EntityTrackerCommunicator communicator = _listeners.get(i);
			communicator.updatedComponentState(entityId, componentIndex, values);
		}
	}

	private RawConnectionCommunicatorProvider _communicatorProvider = new RawConnectionCommunicatorProvider() {
		@Override
		public RawConnectionCommunicator getListener(String remoteName) {
			// Server requests communicator for given remote.

			EntityTrackerCommunicator newCommunicator = new EntityTrackerCommunicator() {
				@Override
				public void connected(SocketAddress remoteAddress, RawConnectionOutputListener output) {
					super.connected(remoteAddress, output);
					injectWorldController(_worldController);


					for (int i = 0, n = _systems.size(); i < n; ++i) {
						final Tuple3<Integer, String, AspectInfo> system = _systems.get(i);
						final AspectInfo aspects = system.item3;
						addedSystem(system.item1, system.item2, aspects.allTypes, aspects.oneTypes, aspects.exclusionTypes);
					}

					for (int i = 0, n = _managers.size(); i < n; ++i) {
						addedManager(_managers.get(i));
					}

					for (int i = 0, n = _componentTypes.size(); i < n; ++i) {
						addedComponentType(i, _componentTypes.get(i));
					}

					for (int i = 0, n = _systems.size(); i < n; ++i) {
						if (_entitySystemsEntitiesCount.get(i) != null) {
							int entitiesCount = _entitySystemsEntitiesCount.get(i);
							int maxEntitiesCount = _entitySystemsMaxEntitiesCount.get(i);
							updatedEntitySystem(i, entitiesCount, maxEntitiesCount);
						}
					}

					for (Entry<Integer, BitVector> entity : _entities.entrySet()) {
						addedEntity(entity.getKey(), entity.getValue());
					}
				}

				@Override
				public void disconnected() {
					_listeners.remove(this);
				}
			};
			_listeners.add(newCommunicator);

			return newCommunicator;
		}
	};
}

File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/network/EntityTrackerServer.kt
package net.namekdev.entity_tracker.network

import java.net.SocketAddress
import java.util.HashMap

import com.artemis.Component
import com.artemis.utils.Bag
import com.artemis.utils.BitVector
import net.namekdev.entity_tracker.connectors.WorldController
import net.namekdev.entity_tracker.connectors.WorldUpdateListener
import net.namekdev.entity_tracker.model.AspectInfo
import net.namekdev.entity_tracker.model.ComponentTypeInfo
import net.namekdev.entity_tracker.network.base.RawConnectionCommunicator
import net.namekdev.entity_tracker.network.base.RawConnectionCommunicatorProvider
import net.namekdev.entity_tracker.network.base.RawConnectionOutputListener
import net.namekdev.entity_tracker.network.base.Server
import net.namekdev.entity_tracker.network.communicator.EntityTrackerCommunicator
import net.namekdev.entity_tracker.utils.tuple.Tuple3


/**
 * Server listening to new clients, useful to pass into Entity Tracker itself.
 * Collects data to gather world state for incoming connections.

 * @author Namek
 */
class EntityTrackerServer @JvmOverloads constructor(listeningPort: Int = Server.DEFAULT_PORT) : Server(), WorldUpdateListener {
    private lateinit var _worldController: WorldController
    private val _listeners = Bag<EntityTrackerCommunicator>()

    private val _managers = Bag<String>()
    private val _systems = Bag<Tuple3<Int, String, AspectInfo>>()
    private val _componentTypes = Bag<ComponentTypeInfo>()
    private val _entities = HashMap<Int, BitVector>()
    private val _entitySystemsEntitiesCount = Bag<Int>()
    private val _entitySystemsMaxEntitiesCount = Bag<Int>()


    override fun injectWorldController(controller: WorldController) {
        _worldController = controller
    }

    override // TODO
    val listeningBitset: Int
        get() = WorldUpdateListener.ENTITY_ADDED or WorldUpdateListener.ENTITY_DELETED or WorldUpdateListener.ENTITY_SYSTEM_STATS

    override fun addedSystem(index: Int, name: String, allTypes: BitVector?, oneTypes: BitVector?, notTypes: BitVector?) {
        var i = 0
        val n = _listeners.size()
        while (i < n) {
            val communicator = _listeners.get(i)
            communicator.addedSystem(index, name, allTypes, oneTypes, notTypes)
            ++i
        }
        _systems.add(Tuple3.create(index, name, AspectInfo(allTypes, oneTypes, notTypes)))
    }

    override fun addedManager(name: String) {
        var i = 0
        val n = _listeners.size()
        while (i < n) {
            val communicator = _listeners.get(i)
            communicator.addedManager(name)
            ++i
        }
        _managers.add(name)
    }

    override fun addedComponentType(index: Int, info: ComponentTypeInfo) {
        var i = 0
        val n = _listeners.size()
        while (i < n) {
            val communicator = _listeners.get(i)
            communicator.addedComponentType(index, info)
            ++i
        }
        _componentTypes.set(index, info)
    }

    override fun updatedEntitySystem(systemIndex: Int, entitiesCount: Int, maxEntitiesCount: Int) {
        var i = 0
        val n = _listeners.size()
        while (i < n) {
            val communicator = _listeners.get(i)
            communicator.updatedEntitySystem(systemIndex, entitiesCount, maxEntitiesCount)
            ++i
        }
        _entitySystemsEntitiesCount.set(systemIndex, entitiesCount)
        _entitySystemsMaxEntitiesCount.set(systemIndex, maxEntitiesCount)
    }

    override fun addedEntity(entityId: Int, components: BitVector) {
        var i = 0
        val n = _listeners.size()
        while (i < n) {
            val communicator = _listeners.get(i)
            communicator.addedEntity(entityId, components)
            ++i
        }
        _entities.put(entityId, components)
    }

    override fun deletedEntity(entityId: Int) {
        var i = 0
        val n = _listeners.size()
        while (i < n) {
            val communicator = _listeners.get(i)
            communicator.deletedEntity(entityId)
            ++i
        }
        _entities.remove(entityId)
    }

    override fun updatedComponentState(entityId: Int, componentIndex: Int, valueTree: Any) {
        // `valueTree` is going to be serialized in next layer
        assert(valueTree is Component)

        var i = 0
        val n = _listeners.size()
        while (i < n) {
            val communicator = _listeners.get(i)
            communicator.updatedComponentState(entityId, componentIndex, valueTree)
            ++i
        }
    }

    private val _communicatorProvider = object : RawConnectionCommunicatorProvider {
        override fun getListener(remoteName: String): RawConnectionCommunicator {
            // Server requests communicator for given remote.

            val newCommunicator = object : EntityTrackerCommunicator() {
                override fun connected(remoteAddress: SocketAddress, output: RawConnectionOutputListener) {
                    super.connected(remoteAddress, output)
                    injectWorldController(_worldController)


                    run {
                        var i = 0
                        val n = _systems.size()
                        while (i < n) {
                            val system = _systems.get(i)
                            val aspects = system.item3
                            addedSystem(system.item1, system.item2, aspects.allTypes, aspects.oneTypes, aspects.exclusionTypes)
                            ++i
                        }
                    }

                    run {
                        var i = 0
                        val n = _managers.size()
                        while (i < n) {
                            addedManager(_managers.get(i))
                            ++i
                        }
                    }

                    run {
                        var i = 0
                        val n = _componentTypes.size()
                        while (i < n) {
                            addedComponentType(i, _componentTypes.get(i))
                            ++i
                        }
                    }

                    var i = 0
                    val n = _systems.size()
                    while (i < n) {
                        if (_entitySystemsEntitiesCount.get(i) != null) {
                            val entitiesCount = _entitySystemsEntitiesCount.get(i)
                            val maxEntitiesCount = _entitySystemsMaxEntitiesCount.get(i)
                            updatedEntitySystem(i, entitiesCount, maxEntitiesCount)
                        }
                        ++i
                    }

                    for ((key, value) in _entities) {
                        addedEntity(key, value)
                    }
                }

                override fun disconnected() {
                    _listeners.remove(this)
                }
            }
            _listeners.add(newCommunicator)

            return newCommunicator
        }
    }

    init {
        super.clientListenerProvider = _communicatorProvider
        super.listeningPort = listeningPort
    }
}
